libcudacxx/include/cuda/std/__cmath/traits.h

```diff
@@ -0,0 +1,469 @@
+{
+#if defined(_CCCL_BUILTIN_ISNAN)
+  return _CCCL_BUILTIN_ISNAN(__x);
+#elif _CCCL_CUDACC_BELOW(11, 8)
+  return ::__isnan(__x);
+#else // ^^^ _CCCL_CUDACC_BELOW(11, 8) ^^^ / vvv !_CCCL_CUDACC_BELOW(11, 8) vvv
+  return ::isnan(__x);
+#endif // !_CCCL_CUDACC_BELOW(11, 8)
+}
+
+#if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI _CCCL_CONSTEXPR_ISNAN bool isnan(long double __x) noexcept
+{
+#  if defined(_CCCL_BUILTIN_ISNAN)
+  return _CCCL_BUILTIN_ISNAN(__x);
+#  elif _CCCL_CUDACC_BELOW(11, 8)
+  return ::__isnan(__x);
+#  else // ^^^ _CCCL_CUDACC_BELOW(11, 8) ^^^ / vvv !_CCCL_CUDACC_BELOW(11, 8) vvv
+  return ::isnan(__x);
+#  endif // !_CCCL_CUDACC_BELOW(11, 8)
+}
+#endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+
+#if defined(_LIBCUDACXX_HAS_NVFP16)
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnan(__half __x) noexcept
+{
+  return ::__hisnan(__x);
+}
+#endif // _LIBCUDACXX_HAS_NVFP16
+
+#if defined(_LIBCUDACXX_HAS_NVBF16)
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnan(__nv_bfloat16 __x) noexcept
+{
+  return ::__hisnan(__x);
+}
+#endif // _LIBCUDACXX_HAS_NVBF16
+
+// isnormal
+
+template <class _A1, enable_if_t<_CCCL_TRAIT(is_integral, _A1), int> = 0>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnormal(_A1 __x) noexcept
+{
+  return __x != 0;
+}
+
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnormal(float __x) noexcept
+{
+  return __x != 0 && _CUDA_VSTD::isfinite(__x);
+}
+
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnormal(double __x) noexcept
+{
+  return __x != 0 && _CUDA_VSTD::isfinite(__x);
+}
+
+#if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnormal(long double __x) noexcept
+{
+  return __x != 0 && _CUDA_VSTD::isfinite(__x);
+}
+#endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+
+#if defined(_LIBCUDACXX_HAS_NVFP16)
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnormal(__half __x) noexcept
+{
+  return _CUDA_VSTD::isnormal(__half2float(__x));
+}
+#endif // _LIBCUDACXX_HAS_NVFP16
+
+#if defined(_LIBCUDACXX_HAS_NVBF16)
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI bool isnormal(__nv_bfloat16 __x) noexcept
+{
+  return _CUDA_VSTD::isnormal(__bfloat162float(__x));
+}
+#endif // _LIBCUDACXX_HAS_NVBF16
+
+// isgreater
+
+template <class _Tp>
+struct __is_extended_arithmetic
+{
+  static constexpr bool value = _CCCL_TRAIT(is_arithmetic, _Tp) || _CCCL_TRAIT(__is_extended_floating_point, _Tp);
+};
+
+#if !defined(_CCCL_NO_VARIABLE_TEMPLATES)
+template <class _Tp>
+_CCCL_INLINE_VAR constexpr bool __is_extended_arithmetic_v =
+  is_arithmetic_v<_Tp> || __is_extended_floating_point_v<_Tp>;
+#endif // !_CCCL_NO_INLINE_VARIABLES
+
+template <class _A1, enable_if_t<_CCCL_TRAIT(__is_extended_arithmetic, _A1), int> = 0>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_HIDE_FROM_ABI bool __device_isgreater(_A1 __x, _A1 __y) noexcept
+{
+  if (_CUDA_VSTD::isnan(__x) || _CUDA_VSTD::isnan(__y))
+  {
+    return false;
+  }
+  return __x > __y;
```

Comment:
Q: Isn't any comparison with `nan` automatically `false` and this could just be:
```suggestion
  return __x > __y;
```
Applies to more places.

Reply:
Yes but that does set the FE_INVALID flag.

The whole purpose of those functions is that they do the comparison without setting that flag. See
https://en.cppreference.com/w/cpp/numeric/math/isgreaterequal